File: 580/C - Kefa and Park/solution.cpp
#include <bits/stdc++.h>
using namespace std;
using i32 = int;
 
class Solution {
public:
  i32 n, m;
  vector<i32> ar;
  vector<vector<i32>> adj;
 
  int dfs(int v, int parent, int cnt) {
    if (cnt > m)
      return 0;
    if ((int)adj[v].size() == 1 && v != 1)
      return 1; // leaf restaurant
 
    int temp = 0;
    for (auto u : adj[v]) {
      if (u == parent)
        continue;
      int nextCnt = (ar[u] == 1) ? cnt + 1 : 0;
      temp += dfs(u, v, nextCnt);
    }
    return temp;
  }
 
  void solve() {
    cin >> n >> m;
    ar.resize(n + 1);
    for (int i = 1; i <= n; i++)
      cin >> ar[i];
 
    adj.resize(n + 1);
    for (int i = 0; i < n - 1; i++) {
      int x, y;
      cin >> x >> y;
      adj[x].push_back(y);
      adj[y].push_back(x);
    }
 
    if (n == 1) {
      cout << 1 << endl;
      return;
    }
 
    int start = (ar[1] == 1) ? 1 : 0;
    cout << dfs(1, 0, start) << endl;
  }
};
 
int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);
  Solution sol;
  sol.solve();
  return 0;
}

File: 580/C - Kefa and Park/solution.rs
use std::io::{self, Read, Write};
 
struct Solution {
    n: usize,
    m: i32,
    ar: Vec<i32>,
    adj: Vec<Vec<usize>>,
}
 
impl Solution {
    fn dfs(&self, v: usize, parent: usize, cnt: i32) -> i64 {
        if cnt > self.m {
            return 0;
        }
        if self.adj[v].len() == 1 && v != 1 {
            return 1; // leaf restaurant
        }
 
        let mut temp: i64 = 0;
        for &u in &self.adj[v] {
            if u == parent {
                continue;
            }
            let next_cnt = if self.ar[u] == 1 { cnt + 1 } else { 0 };
            temp += self.dfs(u, v, next_cnt);
        }
        temp
    }
 
    fn solve(&self) -> i64 {
        if self.n == 1 {
            return 1;
        }
        let start = if self.ar[1] == 1 { 1 } else { 0 };
        self.dfs(1, 0, start)
    }
}
 
fn main() {
    let mut input = String::new();
    io::stdin().read_to_string(&mut input).unwrap();
    let mut it = input
        .split_ascii_whitespace()
        .map(|x| x.parse::<i64>().unwrap());
 
    let n = it.next().unwrap() as usize;
    let m = it.next().unwrap() as i32;
 
    let mut ar = vec![0i32; n + 1];
    for i in 1..=n {
        ar[i] = it.next().unwrap() as i32;
    }
 
    let mut adj = vec![Vec::new(); n + 1];
    for _ in 0..n.saturating_sub(1) {
        let x = it.next().unwrap() as usize;
        let y = it.next().unwrap() as usize;
        adj[x].push(y);
        adj[y].push(x);
    }
 
    let sol = Solution { n, m, ar, adj };
    let result = sol.solve();
 
    let stdout = io::stdout();
    let mut out = stdout.lock();
    writeln!(out, "{}", result).unwrap();
}